src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/MSBuildShellOutNuGetProject.cs

```diff
@@ -176,29 +182,16 @@ private static string GetMSBuildProperty(IVsBuildPropertyStorage buildPropertySt
             return output;
         }
 
-        public PackageSpec GetPackageSpecForRestore()
-        {            
-            var dgSpec = MSBuildUtility.GetProjectReferences(
-                _msbuildPath,
-                new[] { _fullProjectPath },
-                MSBuildWaitTime);
-
-            var packageSpec = dgSpec
-                .Projects
-                .FirstOrDefault(p => _fullProjectPath.Equals(
-                    p.RestoreMetadata.ProjectPath,
-                    StringComparison.OrdinalIgnoreCase));
-
-            // Set the output path, since shelling out to MSBuild does not set this.
-            packageSpec.RestoreMetadata.OutputPath = BaseIntermediateOutputPath;
+        public PackageSpec GetPackageSpecForRestore(ExternalProjectReferenceContext referenceContext)
+        {
+            var specs = GetSpecsAndInitializeInstalledPackages();
```

Comment:
I'll check on this, I think the caller may already have moved off the UI thread.